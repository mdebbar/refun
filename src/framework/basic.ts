// Closure Compiler doesn't like TS enums, so we have to use a class with static
// fields to emulate an enum.
export class DirtyState {
  static Clean = 1;
  static ChildrenOnly = 2;
  static SelfAndChildren = 3;
}
