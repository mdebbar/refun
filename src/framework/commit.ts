import { AppNode, DirtyState, myNode } from './core';
import { SingleUI } from './core/components';
import { trackSkips } from '../instrumentation';

export type Commit = any;
// TODO[types]: This is useful, please bring it back!
export type CommitUI<C extends Committer> = SingleUI;

const commitTracker = trackSkips('commit', 1000);

// A `Node` in the commit tree. It knows how to commit itself during the commit
// phase after the build is complete. In the case of a browser, committing means
// updating html attributes, appending children, applying styles, etc.
export abstract class Committer<D extends Commit = Commit> {
  dirty: DirtyState = DirtyState.Clean;
  private lastCommit: D | null = null;

  needsCommit() {
    this.dirty = DirtyState.SelfAndChildren;
  }

  needsToCheckChildCommits() {
    if (this.dirty === DirtyState.Clean) {
      this.dirty = DirtyState.ChildrenOnly;
    }
  }

  commit(node: CommitNode<this, any>): D {
    if (this.dirty === DirtyState.SelfAndChildren) {
      commitTracker && commitTracker.hit();
    } else {
      commitTracker && commitTracker.skip();
    }

    if (this.dirty === DirtyState.Clean) {
      return this.lastCommit as D;
    }

    const childCommits = commitChildrenRecursive(node);
    const newCommit =
      this.dirty === DirtyState.ChildrenOnly
        ? (this.lastCommit as D)
        : this.commitSelf(this.lastCommit, childCommits);
    this.diffChildren(newCommit, childCommits);
    this.lastCommit = newCommit;
    this.dirty = DirtyState.Clean;
    return newCommit;
  }

  abstract commitSelf(lastCommit: D | null, newChildren: D[]): D;

  abstract diffChildren(newCommit: D, newChildren: D[]): void;
}

export class NoopCommitter extends Committer<void> {
  commitSelf() {}
  diffChildren() {}
}

export class CommitNode<C extends Committer<Commit>, S> extends AppNode<S> {
  constructor(initialState: S, committer: C) {
    super(initialState);
    this.committer = committer;
  }

  committer: C;

  needsChildrenBuild() {
    super.needsChildrenBuild();
    this.committer.needsToCheckChildCommits();
  }

  protected postBuild() {
    super.postBuild();
    this.committer.commit(this);
  }
}

// Commits the children of `node` recursively. As it walks down the tree, it'll
// add children to `arr`.
//
// QUESTION: should we allow commit nodes to throw? e.g. when adding children to
//           an <img>.
function commitChildrenRecursive<D extends Commit>(
  node: AppNode<any> | CommitNode<Committer<D>, any>,
  arr: D[] = [],
): D[] {
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    if (child != null) {
      if (isCommitNode<D>(child)) {
        // When a child is committable, we commit it recursively and add its
        // commit as a child to the ancestor commit.
        const result = child.committer.commit(child);
        if (!result) {
          throw new Error('Expected commit to return something.');
        }
        arr.push(result);
      } else {
        // Continue recursively in the tree to find more commit nodes.
        commitChildrenRecursive(child, arr);
      }
    }
  }
  return arr;
}

function isCommitNode<D>(
  node: AppNode<any>,
): node is CommitNode<Committer<D>, any> {
  return node instanceof CommitNode;
}

export function myCommitNode<C extends Committer<any>>(
  createCommitter: () => C,
) {
  return myNode(() => new CommitNode(null, createCommitter()));
}
