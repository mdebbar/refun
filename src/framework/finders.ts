import { AppNode, Component } from './core';
import { COMPONENT_TYPE } from './core/components';

type Finder = (node: AppNode<any>) => boolean;

export function findAncestor(
  node: AppNode<any>,
  cb: Finder,
): AppNode<any> | null {
  let current: AppNode<any> | null = node;
  while (current != null) {
    if (cb(current)) {
      return current;
    }
    current = current.parent;
  }
  return null;
}

export function findDescendant(node: AppNode<any>, cb: Finder): AppNode<any> {
  const found = silentFindDescendant(node, cb);

  if (found) {
    return found;
  } else {
    throw new Error('Could not find the requested node');
  }
}

export function silentFindDescendant(
  node: AppNode<any>,
  cb: Finder,
): AppNode<any> | null {
  if (cb(node)) {
    return node;
  }

  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i];
    if (child != null) {
      const found = silentFindDescendant(child, cb);
      if (found) return found;
    }
  }

  return null;
}

export function findDescendantOfType(
  node: AppNode<any>,
  component: Component<any>,
): AppNode<any> {
  const type = component()[COMPONENT_TYPE];
  return findDescendant(node, n => n.ui[COMPONENT_TYPE] === type);
}
