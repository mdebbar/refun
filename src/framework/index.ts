export * from './core';
export { committer, CommitNode, Committer, NoopCommitter } from './commit';
export { run } from './run';
// export * from './finders';
export { sleep, animationFrame, animated, gen } from './suspense';
