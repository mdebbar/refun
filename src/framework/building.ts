import { DirtyState } from './basic';
import { commit, CommitNode } from './committing';
import { Render, SingleUI, UI } from './types';
import { SkipRate } from '../instrumentation';

const rate = new SkipRate('build', 0);

// Builds the tree starting at `node` while respecting the `dirtyState` flag.
function build(node: AppNode) {
  switch (node.dirtyState) {
    case DirtyState.Clean:
      // If the node is clean, skip the whole subtree.
      rate.total();
      rate.skip();
      break;

    case DirtyState.ChildrenOnly:
      rate.total();
      rate.skip();
      buildChildren(node);
      break;

    case DirtyState.SelfAndChildren:
      forceBuild(node);
      break;
  }
}

function buildChildren(node: AppNode) {
  const children = node.children;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (child) {
      build(child);
    }
  }
}

function forceBuild(node: AppNode) {
  rate.total();
  const children = node.render(node);
  node.dirtyState = DirtyState.Clean;

  // TODO[perf]: Reuse nodes instead of throwing them away. When a node is
  //             recycled, its state needs to be reset.
  if (Array.isArray(children)) {
    const count = buildMultiUI(children, node, 0);
    if (node.children.length > count) {
      node.children.length = count;
    }
  } else {
    buildUIAtIndex(children, node, 0);
    if (node.children.length > 1) {
      node.children.length = 1;
    }
  }
}

function buildUIAtIndex(ui: SingleUI, parent: AppNode, i: number) {
  if (ui == null) {
    parent.children[i] = null;
  } else {
    const child = prepareChildNode(ui, parent, i);
    forceBuild(child);
  }
}

// Returns how many items it traversed, recursively.
// Because the ui array could be deeply nested, we need to flatten it. But
// doing that on every rebuild is costly. So this function does the same thing
// without the actual flattening.
//
// Because this function could be called recursively while traversing the
// AppNode's children, we need to tell it which child it should start at.
// TODO: handle keys.
function buildMultiUI(multiUI: UI[], parent: AppNode, startAt: number): number {
  let traversed = 0;
  for (let i = 0; i < multiUI.length; i++) {
    const r = multiUI[i];
    if (Array.isArray(r)) {
      // Recursively traverse the nested arrays and keep track of the traversal
      // count.
      traversed += buildMultiUI(r, parent, startAt + traversed);
    } else {
      buildUIAtIndex(r, parent, startAt + i);
      traversed++;
    }
  }
  return traversed;
}

function prepareChildNode(render: Render, parent: AppNode, i: number): AppNode {
  const node = parent.children[i];
  // TODO[perf]: If (node.render === render)
  //             skip building the subtree altogether.
  //             Does this optimization work for generators that need to be
  //             invoked multiple times?
  if (node && isSameComponentType(node.render, render)) {
    // In this case, we should reuse the existing app node.
    node.render = render;
    if (node.commitNode) {
      node.commitNode.dirtyState = DirtyState.SelfAndChildren;
    }
    return node;
  }
  return (parent.children[i] = new AppNode(parent, render));
}

function isSameComponentType(r1: Render, r2: Render) {
  // If both are null we shouldn't consider them of equal types.
  return r1.$$type && r2.$$type && r1.$$type == r2.$$type;
}

// Light-weight tree that maintains the full structure of the rendered component
// tree.
/*

CommitTree        AppTree
    o     <---       o
   / \             /   \
      o   <---    o     o
 / \             / \   / \
o         <---  o   o o   o

*/
export class AppNode<C extends CommitNode = CommitNode> {
  constructor(parent: AppNode | null, render: Render) {
    this.parent = parent;
    this.render = render;
  }

  dirtyState: DirtyState = DirtyState.SelfAndChildren;

  parent: AppNode | null;

  // Some children could be null when their corresponding UI materializes
  // to null. This guarantees that all state is blown away correctly.
  children: (AppNode | null)[] = [];

  // Commit nodes are only required when the component wants to have custom
  // commit logic (e.g. html elements).
  commitNode: C | null = null;

  state: any = null;

  render: Render;

  // Rebuild the sub-tree under this node.
  // TODO: This should accept a callback for updating state (ala React setState).
  // TODO: Make `rebuild` invalid during a build.
  rebuild() {
    this.markAsDirty();
    const rootNode = this.markParentChainAsDirty();
    rootNode.scheduleRebuild();
  }

  alreadyScheduledRebuild = false;

  // Should only be called on the root node.
  scheduleRebuild() {
    if (this.parent != null) {
      throw new Error(
        'scheduleRebuild() should only be called on the root node',
      );
    }
    if (!this.alreadyScheduledRebuild) {
      this.alreadyScheduledRebuild = true;
      requestAnimationFrame(() => {
        this.alreadyScheduledRebuild = false;
        build(this);
        // TODO[types]: This is the root node which is always committable.
        commit(this as CommittableAppNode);
      });
    }
  }

  protected markAsDirty() {
    this.dirtyState = DirtyState.SelfAndChildren;
    if (this.commitNode) {
      this.commitNode.dirtyState = DirtyState.SelfAndChildren;
    }
  }

  // Marks all nodes in the parent chain as dirty and returns the root app node.
  protected markParentChainAsDirty(): AppNode {
    // TODO[perf]: Maybe stop climbing as soon as we find a node that has a
    //             non-Clean dirty state.
    let node: AppNode = this;
    while (node.parent) {
      node = node.parent;
      if (node.dirtyState == DirtyState.Clean) {
        node.dirtyState = DirtyState.ChildrenOnly;
      }
      if (node.commitNode && node.commitNode.dirtyState == DirtyState.Clean) {
        node.commitNode.dirtyState = DirtyState.ChildrenOnly;
      }
    }
    return node;
  }
}

export type CommittableAppNode<C extends CommitNode = CommitNode> = AppNode<
  C
> & {
  commitNode: C;
};

export function isCommittable<C extends CommitNode = CommitNode>(
  node: AppNode<C>,
): node is CommittableAppNode<C> {
  return node.commitNode != null;
}
