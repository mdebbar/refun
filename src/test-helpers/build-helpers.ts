import { AppNode, myStateNode, UI, namedComponent } from '../framework';

export function createRef(name: string) {
  return namedComponent(name, (children: () => UI) => {
    myStateNode(null);
    return children();
  });
}

export function manageFrames(rootNode: AppNode<any>): VoidFunction {
  let _cb: VoidFunction | null = null;
  rootNode.scheduleNextFrame = cb => {
    _cb = cb;
  };

  return function nextFrame() {
    if (_cb == null) {
      throw new Error('There was no scheduled rebuild');
    }
    _cb();
    _cb = null;
  };
}
