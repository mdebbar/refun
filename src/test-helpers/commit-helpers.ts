import { Committer, myCommitNode, component, UI } from '../framework';

export class TestCommit {
  // Can be used to check the order of commit creation and application.
  value: string;
  children: TestCommit[];

  constructor(value: string) {
    this.value = value;
    this.children = [];
  }
}

export class TestCommitter extends Committer<TestCommit> {
  private data: string;
  private testCommit: TestCommit | null = null;

  constructor(data: string) {
    super();
    this.data = data;
  }

  getTestCommit(): TestCommit {
    return this.testCommit as TestCommit;
  }

  update(newData: string) {
    this.data = newData;
    this.needsCommit();
  }

  commitSelf(lastCommit: TestCommit, newChildren: TestCommit[]): TestCommit {
    if (!lastCommit) {
      return new TestCommit(this.data);
    }
    lastCommit.value = `${lastCommit.value}|${this.data}`;
    return lastCommit;
  }

  diffChildren(newCommit: TestCommit, newChildren: TestCommit[]): void {
    newCommit.value = newCommit.value + '*';
    this.testCommit = newCommit;
    newCommit.children = newChildren;
  }
}

export const committer = component(function committer(
  str: string,
  children: UI,
) {
  const node = myCommitNode(() => new TestCommitter(str));
  node.committer.update(str);
  return children;
});
