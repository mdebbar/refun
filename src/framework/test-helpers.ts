import { Committer } from './commit';

let incrementalCommitValue = 0;

// This function has to be called at the beginning of any test that uses
// jest snapshots with commits.
export function resetCommitValue() {
  incrementalCommitValue = 0;
}

export class TestCommit {
  // Can be used to check the order of commit creation and application.
  value: number;
  children: TestCommit[];

  constructor(value: number) {
    this.value = value;
    this.children = [];
  }
}

export class TestCommitter extends Committer<TestCommit> {
  commitSelf(lastCommit: TestCommit, newChildren: TestCommit[]): TestCommit {
    if (!lastCommit) {
      return new TestCommit(incrementalCommitValue++);
    }
    lastCommit.value = incrementalCommitValue++;
    return lastCommit;
  }

  diffChildren(newCommit: TestCommit, newChildren: TestCommit[]): void {
    newCommit.children = newChildren;
  }
}
