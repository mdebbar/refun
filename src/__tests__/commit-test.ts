import { run, AnyUI, component, myStateNode } from '../framework';
import { TestCommitter, committer } from '../test-helpers/commit-helpers';
import {
  stateSerializer,
  commitSerializer,
} from '../test-helpers/snapshot-serializers';
import { findDescendantOfType } from '../framework/finders';
import { createRef, manageFrames } from '../test-helpers/build-helpers';

expect.addSnapshotSerializer(stateSerializer);
expect.addSnapshotSerializer(commitSerializer);

const app = (children: AnyUI) => children;

const stateless1 = (children: AnyUI) => children;
const stateless2 = (children: AnyUI) => stateless1(children);
const statelessCommitter = (str: string, children: AnyUI) =>
  stateless1(committer(str, children));

const stateful = component(function stateful(children: AnyUI) {
  const node = myStateNode(0);
  node.state++;
  return stateless2(children);
});

const statefulCommitter = component(function statefulCommitter(
  str: string,
  children: AnyUI,
) {
  const node = myStateNode(0);
  node.state++;
  return stateless2(committer(str, children));
});

test('simple commit', () => {
  const rootNode = run(() => app(committer('1', null)), new TestCommitter('0'));
  expect(rootNode.committer.getTestCommit()).toMatchSnapshot();
});

test('complex commits', () => {
  const complex = () =>
    app([
      committer('1', stateless2(committer('2', null))),
      stateless2([
        committer('3', [stateful(null)]),
        null,
        statefulCommitter('4', [
          stateless1(stateful([statefulCommitter('5', null)])),
          stateful([]),
          committer('6', null),
          stateless2(null),
        ]),
      ]),
      [stateless1(statelessCommitter('7', []))],
    ]);
  const rootNode = run(() => complex(), new TestCommitter('0'));
  expect(rootNode).toMatchSnapshot();
  expect(rootNode.committer.getTestCommit()).toMatchSnapshot();
});

test('simple recommit', () => {
  const root = () => stateless1(statefulCommitter('1', null));
  const rootNode = run(root, new TestCommitter('0'));
  const nextFrame = manageFrames(rootNode);

  expect(rootNode).toMatchSnapshot('before rebuild');
  expect(rootNode.committer.getTestCommit()).toMatchSnapshot('before rebuild');

  rootNode.rebuild();
  nextFrame();
  expect(rootNode).toMatchSnapshot('after rebuild');
  expect(rootNode.committer.getTestCommit()).toMatchSnapshot('after rebuild');
});

test('complex recommit', () => {
  const ref1 = createRef('ref1');
  const ref2 = createRef('ref2');

  const complex = () =>
    app([
      committer('1', stateless2(committer('2', null))),
      ref2(() =>
        stateless2([
          committer('3', [stateful(null)]),
          null,
          statefulCommitter('4', [
            ref1(() => stateless1(stateful([statefulCommitter('5', null)]))),
            stateful([]),
            committer('6', null),
            stateless2(null),
          ]),
        ]),
      ),
      [stateless1(statelessCommitter('7', []))],
    ]);

  const flat = () =>
    app([
      committer('1', null),
      ref2(() => committer('3', [stateful(null)])),
      null,
      statefulCommitter('4', [
        [statefulCommitter('5', null)],
        stateful([]),
        committer('6', null),
        stateless2(null),
      ]),
      statelessCommitter('7', []),
    ]);

  let current = complex;
  const rootNode = run(() => current(), new TestCommitter('0'));
  const nextFrame = manageFrames(rootNode);

  expect(rootNode).toMatchSnapshot('before rebuild');
  expect(rootNode.committer.getTestCommit()).toMatchSnapshot('before rebuild');

  findDescendantOfType(rootNode, ref1).rebuild();
  findDescendantOfType(rootNode, ref2).rebuild();
  nextFrame();
  expect(rootNode).toMatchSnapshot('after partial rebuild');
  expect(rootNode.committer.getTestCommit()).toMatchSnapshot(
    'after partial rebuild',
  );

  current = flat;
  rootNode.rebuild();
  nextFrame();
  expect(rootNode).toMatchSnapshot('after rebuild to flat');
  expect(rootNode.committer.getTestCommit()).toMatchSnapshot(
    'after rebuild to flat',
  );
});
