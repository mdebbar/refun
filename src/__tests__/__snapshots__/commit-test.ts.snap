// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex commits 1`] = `
%root%
null
commit value: 0*
[
  %committer%
  null
  commit value: 1*
  [
    %committer%
    null
    commit value: 2*
    [
      null
    ]
  ],
  %committer%
  null
  commit value: 3*
  [
    %stateful%
    1
    [
      null
    ]
  ],
  null,
  %statefulCommitter%
  1
  [
    %committer%
    null
    commit value: 4*
    [
      %stateful%
      1
      [
        %statefulCommitter%
        1
        [
          %committer%
          null
          commit value: 5*
          [
            null
          ]
        ]
      ],
      %stateful%
      1
      [],
      %committer%
      null
      commit value: 6*
      [
        null
      ],
      null
    ]
  ],
  %committer%
  null
  commit value: 7*
  []
]
`;

exports[`complex commits 2`] = `
#0* [
  #1* [
    #2* []
  ],
  #3* [],
  #4* [
    #5* [],
    #6* []
  ],
  #7* []
]
`;

exports[`complex recommit: after partial rebuild 1`] = `
%root%
null
commit value: 0**
[
  %committer%
  null
  commit value: 1*
  [
    %committer%
    null
    commit value: 2*
    [
      null
    ]
  ],
  %ref2%
  null
  [
    %committer%
    null
    commit value: 3*|3*
    [
      %stateful%
      2
      [
        null
      ]
    ],
    null,
    %statefulCommitter%
    2
    [
      %committer%
      null
      commit value: 4*|4*
      [
        %ref1%
        null
        [
          %stateful%
          2
          [
            %statefulCommitter%
            2
            [
              %committer%
              null
              commit value: 5*|5*
              [
                null
              ]
            ]
          ]
        ],
        %stateful%
        2
        [],
        %committer%
        null
        commit value: 6*|6*
        [
          null
        ],
        null
      ]
    ]
  ],
  %committer%
  null
  commit value: 7*
  []
]
`;

exports[`complex recommit: after partial rebuild 2`] = `
#0** [
  #1* [
    #2* []
  ],
  #3*|3* [],
  #4*|4* [
    #5*|5* [],
    #6*|6* []
  ],
  #7* []
]
`;

exports[`complex recommit: after rebuild to flat 1`] = `
%root%
null
commit value: 0***
[
  %committer%
  null
  commit value: 1*|1*
  [
    null
  ],
  %ref2%
  null
  [
    %committer%
    null
    commit value: 3*|3*|3*
    [
      %stateful%
      3
      [
        null
      ]
    ]
  ],
  null,
  %statefulCommitter%
  1
  [
    %committer%
    null
    commit value: 4*
    [
      %statefulCommitter%
      1
      [
        %committer%
        null
        commit value: 5*
        [
          null
        ]
      ],
      %stateful%
      1
      [],
      %committer%
      null
      commit value: 6*
      [
        null
      ],
      null
    ]
  ],
  %committer%
  null
  commit value: 7*
  []
]
`;

exports[`complex recommit: after rebuild to flat 2`] = `
#0*** [
  #1*|1* [],
  #3*|3*|3* [],
  #4* [
    #5* [],
    #6* []
  ],
  #7* []
]
`;

exports[`complex recommit: before rebuild 1`] = `
%root%
null
commit value: 0*
[
  %committer%
  null
  commit value: 1*
  [
    %committer%
    null
    commit value: 2*
    [
      null
    ]
  ],
  %ref2%
  null
  [
    %committer%
    null
    commit value: 3*
    [
      %stateful%
      1
      [
        null
      ]
    ],
    null,
    %statefulCommitter%
    1
    [
      %committer%
      null
      commit value: 4*
      [
        %ref1%
        null
        [
          %stateful%
          1
          [
            %statefulCommitter%
            1
            [
              %committer%
              null
              commit value: 5*
              [
                null
              ]
            ]
          ]
        ],
        %stateful%
        1
        [],
        %committer%
        null
        commit value: 6*
        [
          null
        ],
        null
      ]
    ]
  ],
  %committer%
  null
  commit value: 7*
  []
]
`;

exports[`complex recommit: before rebuild 2`] = `
#0* [
  #1* [
    #2* []
  ],
  #3* [],
  #4* [
    #5* [],
    #6* []
  ],
  #7* []
]
`;

exports[`simple commit 1`] = `
#0* [
  #1* []
]
`;

exports[`simple recommit: after rebuild 1`] = `
%root%
null
commit value: 0**
[
  %statefulCommitter%
  2
  [
    %committer%
    null
    commit value: 1*|1*
    [
      null
    ]
  ]
]
`;

exports[`simple recommit: after rebuild 2`] = `
#0** [
  #1*|1* []
]
`;

exports[`simple recommit: before rebuild 1`] = `
%root%
null
commit value: 0*
[
  %statefulCommitter%
  1
  [
    %committer%
    null
    commit value: 1*
    [
      null
    ]
  ]
]
`;

exports[`simple recommit: before rebuild 2`] = `
#0* [
  #1* []
]
`;
